### PR TITLE
Specifications for Dapr Binding Implementation

### PR DESCRIPTION
**Pull request by SoTrx**
Hey,

Here are the proposed specifications for the Dapr Binding implementation.

### CODE REVIEW

File: resources/2024-10-dapr-bindings.md
Comment (lakshmimsft) [inline resources/2024-10-dapr-bindings.md:1]: missing title.

File: resources/2024-10-dapr-bindings.md
Comment (lakshmimsft) [inline resources/2024-10-dapr-bindings.md:7]: The first sentence may raise more questions rather than help provide an overview for the design. Adding Dapr Bindings will be great but will not complete support for Dapr. Maybe we avoid the first sentence and add more information on how adding Dapr Bindings will ,say, help resources interact with external systems in a decoupled manner.

File: resources/2024-10-dapr-bindings.md
Comment (lakshmimsft) [inline resources/2024-10-dapr-bindings.md:185]: 'frontend' doesn't seem right. Suggestion: Extending the TypeSpec schema definition to include the new Application.Dapr/Bindings resource

File: resources/2024-10-dapr-bindings.md
Comment (lakshmimsft) [inline resources/2024-10-dapr-bindings.md:195]: Can you pls elaborate more.

File: resources/2024-10-dapr-bindings.md
Comment (lakshmimsft) [inline resources/2024-10-dapr-bindings.md:230]: we should rename 'Dapr RP' everywhere in the document to 'Applications.Dapr RP/resource provider'

File: resources/2024-10-dapr-bindings.md
Comment (lakshmimsft) [inline resources/2024-10-dapr-bindings.md:242]: Can we describe briefly how the system will handle errors when communicating with the external systems in case  they are down or unresponsive?

File: resources/2024-10-dapr-bindings.md
Comment (lakshmimsft) [inline resources/2024-10-dapr-bindings.md:255]: We may need some details here on how resources can/should connect and communicate securely with external resources with, say, using secret stores.

File: resources/2024-10-dapr-bindings.md
Comment (lakshmimsft) [inline resources/2024-10-dapr-bindings.md:177]: Thanks a lot for the diagram! If possible pls add more details on calls eg. input versus output binding request

File: resources/2024-10-dapr-bindings.md
Comment (lakshmimsft) [inline resources/2024-10-dapr-bindings.md:227]: 👍

File: resources/2024-10-dapr-bindings.md
Comment (lakshmimsft) [inline resources/2024-10-dapr-bindings.md:72]: 👍 thanks for the detailed code snippets and examples!

**Comment by lakshmimsft** [comment]
Hello @SoTrx. Big thanks for the design doc PR. I've given some initial comments mainly requesting more details in some areas but looking good overall. Thanks a lot!

File: resources/2024-10-dapr-bindings.md
Comment (lakshmimsft) [inline resources/2024-10-dapr-bindings.md:31]: Let's discuss this in the design meeting. I can imagine that recipes may be valuable here.

File: resources/2024-10-dapr-bindings.md
Comment (SoTrx) [reply resources/2024-10-dapr-bindings.md:286]: I've added some information on that topic, but it may be out of scope as it describes how the Dapr sidecar reacts.

File: resources/2024-10-dapr-bindings.md
Comment (SoTrx) [reply resources/2024-10-dapr-bindings.md:31]: Sure !

**Review by rynowak — COMMENTED** [review]
Hey @SoTrx - just wanted to drop by and say thanks for doing this. I'll be traveling during the discussion on Monday. I think the template is well filled out 👍 and will lead to the right discussion.

File: resources/2024-10-dapr-bindings.md
Comment (willtsai) [inline resources/2024-10-dapr-bindings.md:33]: I think these focus on user scenarios for the Dapr Bindings themselves, which is helpful, but I'd recommend adding some scenarios on how a user would use Radius to bring a Dapr Binding into their application

**Review by kachawla — COMMENTED** [review]
Thanks @SoTrx, this is a well written doc! Main area that I would like to discuss is the proposal to support manual provisioning instead of recipe based provisioning, pros and cons that led to this decision.

**Comment by ytimocin** [comment]
> Thanks @SoTrx, this is a well written doc! Main area that I would like to discuss is the proposal to support manual provisioning instead of recipe based provisioning, pros and cons that led to this decision.

Just to better understand @SoTrx, we are going to be supporting both types of provisioning by default, right?

File: resources/2024-10-dapr-bindings.md
Comment (SoTrx) [reply resources/2024-10-dapr-bindings.md:30]: (Recap meeting)
Typically, there are two entities to manage with a Dapr building block: the Dapr CRD and the associated resource it connects to.

Bindings are unique because the resource they connect to is external to the deployed application. Therefore, bindings only create the CRD and do not manage or provision the external resource.

The highlighted sentence will be corrected to reflect this distinction

File: resources/2024-10-dapr-bindings.md
Comment (SoTrx) [reply resources/2024-10-dapr-bindings.md:94]: (Recap meeting)
Each binding has a direction metadata, indicating whether it is an output binding or an input binding. Some bindings can function as both input and output bindings.

File: resources/2024-10-dapr-bindings.md
Comment (SoTrx) [reply resources/2024-10-dapr-bindings.md:161]: (Recap meeting)
A Dapr subscription is indeed another CRD, used to create what's called a declarative subscription.

However, this is not what's reflected in the mermaid diagram. In this case, "subscription" refers to the Dapr sidecar using its integrated SDKs to establish a connection to an external resource. Depending on the resource being listened to, this connection could involve long polling or a subscription to a topic (e.g., subscribing to a RabbitMQ queue using the SDK).

This will be corrected in the diagram, as it is misleading

File: resources/2024-10-dapr-bindings.md
Comment (SoTrx) [reply resources/2024-10-dapr-bindings.md:238]: (Recap meeting)
My proposal is to wait either until all the building blocks are implemented to do some refactoring or to wait for User Defined Types to assess the consequences.

File: resources/2024-10-dapr-bindings.md
Comment (SoTrx) [reply resources/2024-10-dapr-bindings.md:33]: (Recap meeting)
The user story will be updated to reflect that

**Comment by SoTrx** [comment]
> > Thanks @SoTrx, this is a well written doc! Main area that I would like to discuss is the proposal to support manual provisioning instead of recipe based provisioning, pros and cons that led to this decision.
> 
> Just to better understand @SoTrx, we are going to be supporting both types of provisioning by default, right?

In my view, recipes aren't necessary for bindings, as they only create the CRD and could lead to a two-step process for creating a single resource.

However, since this indeed contradicts Radius' design policy, how about introducing "common usage binding recipes"?

For instance, we could define an "object-storage binding recipe" that:

    In local development: creates both the binding CRD and the storage backend (e.g., MinIO).
    In production (Azure): only creates the binding CRD, with a parameter in the recipe to provide the storage account details.

This approach aligns with Radius' philosophy, is convenient for developers, and retains the core concept of bindings.

File: resources/2024-10-dapr-bindings.md
Comment (SoTrx) [reply resources/2024-10-dapr-bindings.md:110]: (Recap meeting)
Yes, Dapr is required for bindings to work. Behind the scenes, a binding instructs the Dapr sidecar to perform a specific type of action

File: resources/2024-10-dapr-bindings.md
Comment (SoTrx) [reply resources/2024-10-dapr-bindings.md:174]: (Recap meeting)
It's the CRD !
I'll update the graph to represent this

File: resources/2024-10-dapr-bindings.md
Comment (rynowak) [inline resources/2024-10-dapr-bindings.md:39]: I think it's also fair to say that this is *in-scope* when the user writes recipes. Writing or choosing the recipes will be user's responsibility.

File: resources/2024-10-dapr-bindings.md
Comment (rynowak) [inline resources/2024-10-dapr-bindings.md:233]: This is a valid usecase (pre-existing resource).

We'd usually think about the platform or infrastructure team as being the one to choose the recipes for production usage.

File: resources/2024-10-dapr-bindings.md
Comment (willtsai) [inline resources/2024-10-dapr-bindings.md:41]: In both user scenarios, I recommend elaborating on the user ultimately leveraging Dapr Bindings to accomplish the abstraction, and then they are wanting to make use of these Dapr Bindings in Radius applications. e.g.
> As an operator using Radius, I want the application I manage to interface with other applications in my organization without exposing their implementation details to the developers by leveraging Dapr Bindings. Thus, I would like to define and use Dapr Bindings as a resource type within my Radius applications.

File: resources/2024-10-dapr-bindings.md
Comment (rynowak) [inline resources/2024-10-dapr-bindings.md:286]: If the type name is `bindings` then this should be `TrackedResourceRequired<DaprBindingProperties, "bindings">`

File: resources/2024-10-dapr-bindings.md
Comment (rynowak) [reply resources/2024-10-dapr-bindings.md:249]: Resource type names are `camelCase`.

File: resources/2024-10-dapr-bindings.md
Comment (rynowak) [inline resources/2024-10-dapr-bindings.md:368]: I don't quite understand this - apologies, I would out of the office when this was discussed. We have existing recipes for Dapr components that create both the infrastructure and the Dapr `Component` resource, so I'm not aware of any issues. LMK if I'm missing anything!

**Review by ytimocin — APPROVED** [review]
Other than a few nits, LGTM. Thanks!

**Review by rynowak — APPROVED** [review]
LGTM - thanks for the updates!

File: resources/2024-10-dapr-bindings.md
Comment (SoTrx) [reply resources/2024-10-dapr-bindings.md:368]: Sorry for the delay!

There is no issue with the creation process itself. You can create both the Binding CRD and the infrastructure just as you would with any other Dapr Building Block. Users can (and likely will) do this with their own configurations, and we certainly shouldn’t try to restrict that flexibility.

My point is only relevant to the initial binding recipes we plan to provide to users. Here’s some context: the Bindings Building Block is unique in Dapr. It predates most other building blocks and originally served as a way to accomplish tasks that did not yet have dedicated building blocks. This is why some bindings overlap with other blocks, such as Azure Storage Queue.

Bindings are generally basic, often supporting only CRUD-like operations. Additionally, Bindings is the only building block without a unified feature set across multiple backends—nearly every binding is different, and few are interchangeable.

Today, bindings are intended for interacting with external resources, as they are designed for basic interactions. For internal, app-scoped backing services, there is usually a better building block available.

While this approach doesn’t affect how Radius implements Bindings, it does impact the default recipes we include in radius/recipes. If we consider bindings best suited for external resources, then a binding recipe should only contain the Dapr Component CRD. This is why I initially opposed creating binding recipes, as it would involve a two-step process for what is effectively a single CRD creation.

In the first meeting, however, a very good point was raised: from Radius's perspective, it is difficult to justify not having any recipes for a resource.

So, my revised proposal is to create default recipes with this "external resource" perspective in mind. These recipes would set up the backing service in local-dev for convenience, but in production, they would have as input the credentials to an existing backing service. This approach isn’t meant to restrict users but rather as a way to nudge users in the right direction from the start. Of course, users can create custom recipes to override this behavior as they see fit.

I'm not entirely sure if I explained this clearly or if it make sense, so please don’t hesitate to challenge me or ask for clarification on any part of this! And if this direction makes sense, I’m happy to revise the design doc to reflect it.

Comment (rynowak) [reply resources/2024-10-dapr-bindings.md:368]: > So, my revised proposal is to create default recipes with this "external resource" perspective in mind. These recipes would set up the backing service in local-dev for convenience, but in production, they would have as input the credentials to an existing backing service. This approach isn’t meant to restrict users but rather as a way to nudge users in the right direction from the start. Of course, users can create custom recipes to override this behavior as they see fit.

Hey @SoTrx - I appreciate the response 👍 

I think the part where I'm stuck is why we need **recipes** for the "pre-existing" case. This is what `resourceProvisioning: 'manual'` is for.... when you want to configure the metadata directly.  

I'm happy to continue this discussion when we start looking at recipe examples since it's not really going to change the design of the `bindings` resource type.

**Comment by rynowak** [comment]
Hi @SoTrx - can you take a look at the spellcheck errors? https://github.com/radius-project/design-notes/actions/runs/11671732590/job/32742154388?pr=68

For any false-positives reported by the spellcheck you can add them here: https://github.com/radius-project/design-notes/blob/main/.github/config/en-custom.txt

Or if they are technical terms I like to use backquotes to turn them into code. ``like this!``

After that I think this is ready to merge.

**Comment by SoTrx** [comment]
> Hi @SoTrx - can you take a look at the spellcheck errors? https://github.com/radius-project/design-notes/actions/runs/11671732590/job/32742154388?pr=68
> 
> For any false-positives reported by the spellcheck you can add them here: https://github.com/radius-project/design-notes/blob/main/.github/config/en-custom.txt
> 
> Or if they are technical terms I like to use backquotes to turn them into code. `like this!`
> 
> After that I think this is ready to merge.
Hey Ryan, 
Sorry, I haven’t been able to work on this lately. Could you rerun the CI? I can’t view the latest spellcheck summary.
Thanks!

**Review by rynowak — APPROVED** [review]

**Comment by rynowak** [comment]
Hi @SoTrx - there's a few more spelling errors here: https://github.com/radius-project/design-notes/actions/runs/12241896918/job/34148057351

Can you take a look?

**Comment by rynowak** [comment]
Looks like the merge conflicts keep coming. Would be it be helpful for one of us maintainers to help finalize this? I know it's hard when things keep moving fast.

**Review by rynowak — APPROVED** [review]

**Comment by rynowak** [comment]
Merging!

**Pull request merged**
